### PR TITLE
Let `BeanStateReader` handle generated types transparently

### PR DESCRIPTION
**Pull request by bamboo**
So generated classes can be encoded transparently.

### Reviewing cheatsheet

Before merging the PR, comments starting with 
- ❌ ❓**must** be fixed
- 🤔 💅 **should** be fixed
- 💭 **may** be fixed
- 🎉 celebrate happy things

### CODE REVIEW

**Comment by bamboo** [comment]
@bot-gradle test ACC

File: platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultClassEncoder.kt [hunk 1]
```diff
@@ -76,9 +77,12 @@ class DefaultClassEncoder(
         } else {
             val newId = classes.putInstance(type)
             writeSmallInt(newId)
-            val className = type.name
+            // TODO:configuration-cache - should collect the details of the decoration (eg enabled annotations, etc), and also carry this information with the serialized class reference
+            val originalType = GeneratedSubclasses.unpack(type)
```
Comment (bamboo) [inline platforms/core-configuration/configuration-cache/src/main/kotlin/org/gradle/internal/cc/impl/serialize/DefaultClassEncoder.kt:81]: This is the main change. The information whether the type is generated is encoded with the type and upon decoding, generated types are reconstructed.

File: platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/BeanPropertyExtensions.kt [hunk 2]
```diff
@@ -28,7 +28,7 @@ import java.lang.reflect.InvocationTargetException
  *
  * A property can only be written when there's a suitable [Codec] for its [value].
  */
-suspend fun WriteContext.writeNextProperty(name: String, value: Any?, kind: PropertyKind) {
+suspend fun WriteContext.writePropertyValue(kind: PropertyKind, name: String, value: Any?) {
```
Comment (bamboo) [inline platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/BeanPropertyExtensions.kt:31]: 💅 Just a minor refactoring for symmetry.

File: platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/Codec.kt [hunk 3]
```diff
@@ -359,12 +359,12 @@ interface BeanStateWriter {
...
 
-    fun ReadContext.newBeanWithId(generated: Boolean, id: Int) =
-        newBean(generated).also {
+    fun ReadContext.newBeanWithId(id: Int) =
+        newBean().also {
             isolate.identities.putInstance(id, it)
         }
 
-    fun ReadContext.newBean(generated: Boolean): Any
+    fun ReadContext.newBean(): Any
```
Comment (bamboo) [inline platforms/core-configuration/graph-serialization/src/main/kotlin/org/gradle/internal/serialize/graph/Codec.kt:367]: As a consequence of the main change, clients no longer need to keep track whether specific types are generated or not.

File: platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AbstractClassGenerator.java [hunk 4]
```diff
@@ -178,23 +179,15 @@ PropertyRoleAnnotationHandler getRoleHandler() {
...
-        GeneratedClassImpl generatedClass = generatedClasses.getIfPresent(type);
-        if (generatedClass == null) {
-            // It is possible that multiple threads will execute this branch concurrently, when the type is missing. However, the contract for `get()` below will ensure that
-            // only one thread will actually generate the implementation class
-            generatedClass = generatedClasses.get(type, generator);
-            // Also use the generated class for itself
-            generatedClasses.put(generatedClass.generatedClass, generatedClass);
-        }
-        return Cast.uncheckedNonnullCast(generatedClass);
+        return Cast.uncheckedNonnullCast(generatedClasses.get(unpack(type), generator));
```
Comment (bamboo) [inline platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/generator/AbstractClassGenerator.java:190]: 🐎 There was a race condition between `generatedClasses.get(...)` and `generatedClasses.put(generatedClass.generatedClass, ...)` that could cause generated types to be regenerated (e.g. `MyType_Decorated_Decorated`).

**Comment by bamboo** [comment]
@bot-gradle test

File: platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/DeserializationInstantiator.java [hunk 5]
```diff
@@ -28,4 +28,9 @@ public interface DeserializationInstantiator {
      * @throws ObjectInstantiationException On failure to create the new instance.
      */
     <T> T newInstance(Class<T> implType, Class<? super T> baseClass) throws ObjectInstantiationException;
+
+    /**
+     * Returns the generated type for the given implementation type.
+     */
+    <T> Class<? extends T> getDeserializedType(Class<T> implType);
```
Comment (abstratt) [inline platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/DeserializationInstantiator.java:35]: Wouldn't `getGeneratedType(...)` be more accurate?

Comment (bamboo) [reply platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/DeserializationInstantiator.java:35]: Sounds right 😄

Comment (bamboo) [reply platforms/core-configuration/model-core/src/main/java/org/gradle/internal/instantiation/DeserializationInstantiator.java]: Renamed.

**Review by abstratt — COMMENTED** [review]
This is a much needed improvement, thanks!

**Pull request merged**